tests/unit/test_main.py

```diff
@@ -1016,3 +1016,582 @@ def test_match_and_apply(self, flow, args, resp):
+                            "port": 17
+                        }]}
+                    ],
+                    "table_id": 2,
+                    "table_group": "base",
+                    "priority": 20000,
+                    }, },
+                {"flow": {
+                    "match": {
+                        "in_port": 20,
+                        "dl_vlan": 201
+                        },
+                    "instructions": [{
+                        "instruction_type": "apply_actions",
+                        "actions": [{
+                            "action_type": "send_report"
+                        }]},
+                        {
+                            "instruction_type": "goto_table",
+                            "table_id": 2
+                        }
+                    ],
+                    "table_id": 0,
+                    "table_group": "evpl",
+                    "priority": 20000,
+                    }, },
+                {"flow": {
+                    "match": {
+                        "in_port": 20,
+                        "dl_vlan": 201
+                        },
+                    "instructions": [{
+                        "instruction_type": "apply_actions",
+                        "actions": [{
+                            "action_type": "pop_int"
+                        }, {
+                            "action_type": "set_vlan",
+                            "vlan_id": 200
+                        }, {
+                            "action_type": "output",
+                            "port": 16
+                        }]}
+                    ],
+                    "table_id": 2,
+                    "table_group": "base",
+                    "priority": 20000,
+                    }, },
+                ]}
+        payload = [
+            {"trace": {
+                "switch": {
+                    "dpid": "00:00:00:00:00:00:00:01",
+                    "in_port": 15
+                    },
+                "eth": {"dl_vlan": 201}
+                }},
+            {"trace": {
+                "switch": {
+                    "dpid": "00:00:00:00:00:00:00:01",
+                    "in_port": 16
+                    },
+                "eth": {"dl_vlan": 200}
+                }},
+            {"trace": {
+                "switch": {
+                    "dpid": "00:00:00:00:00:00:00:01",
+                    "in_port": 20
+                    },
+                "eth": {"dl_vlan": 201}
+                }}
+        ]
+
+        resp = await self.api_client.put(self.traces_endpoint, json=payload)
+        assert resp.status_code == 200
+        current_data = resp.json()
+        result = current_data["result"]
+
+        result_ = result[0]
+        assert result_[0]['dpid'] == '00:00:00:00:00:00:00:01'
+        assert result_[0]['port'] == 15
+        assert result_[0]['type'] == 'last'
+        assert result_[0]['vlan'] == 201
+        assert result_[0]['out']['port'] == 19
```

Comment:
![20231006_144811](https://github.com/kytos-ng/sdntrace_cp/assets/1010796/33ebecfe-1da7-44c1-a4e8-4d9f533833ac)

@gretelliz the initial part is correct, but it's not complete. Here's the thing with an internal loop, when an outgoing port has a link to the same switch it should still continue to process. Notice that port 19-20 has an "internal loop" a link between them, this also needs to be mocked and the trace should be able to trace correctly. The trace termination for this payload should be on port 16:

```
 {"trace": {
                "switch": {
                    "dpid": "00:00:00:00:00:00:00:01",
                    "in_port": 15
                    },
                "eth": {"dl_vlan": 201}
                }
  ```

Reply:
> has a link to the same switch it should still continue to process.

To clarify what I mean by continue to process, I mean it should continue the trace steps while still concluding an outgoing port as a step. However, this internal loop based on how flows are installed it's unidirectional it should not be categorized as loop (which looks good since how you implemented [has_loop](https://github.com/kytos-ng/sdntrace_cp/blob/master/main.py#L157) it needs to repeat a step entering in the same dpid in the same incoming port, so in this case it won't be categorized as a loop.

Reply:
Ok.